### PR TITLE
feat(robot-server): create middleware to support api versioning

### LINKED ISSUE #4633: Robot HTTP API: Establish versioning
## overview

Step one of refactoring the HTTP API while not breaking backwards compatibility is to establish versioning.

HTTP API versions are single integers that always increase.

~The API server and update server should each have (separate) HTTP API versions. These are minimum and maximum versions.~

## new behavior

~These versions should be~
- ~Expressed in the health endpoints:~
- captured from inputs
~`application/vnd.opentrons.api.${version}` (or `update.${version}`) should be captured from the `Accept` header of requests. If no `Accept` header is passed, the version is assumed to be 1.~

- expressed in outputs
~All responses should have the header `X-Opentrons-Media-Type: opentrons.api.${version}` (or `update.${version}`~

~The server should respond to unsupported version requests from the `Accept` header with `406 Not Acceptable` and an error message like this:~

---

@mcous edit on 2020-04-16 now that #5438 is open:

- Requests may include a header that specifies the HTTP API version to use
    - That header should be `Opentrons-Version`
    - If this header is missing, it must be assumed that the client is asking for the latest version the server supports
- Responses must include a header that specifies the HTTP API version the server actually used to process the request
    - That header should also be `Opentrons-Version`

Reasoning and background:

- `X-...` headers have [been deprecated](https://tools.ietf.org/html/rfc6648) and are not recommended for new APIs
    - Prefixing any headers with `Opentrons-` should be sufficient for our needs
    - Header value stolen from the [Stripe API](https://stripe.com/docs/api)
- Overloading the `Accept` header with version conflicts with various specs and seems annoying from a string-parsing standpoint
    - Let's report utf-8 JSON, because that's really what we accept and version isn't _really_ related to media type (depending on who you ask)
- Minimum supported API versions an HTTP API are really hard to manage
     - If we ever need to make a big enough breaking change to the API that they can't be handled with version headers, we need to change paths
     - For now that means no concept of "minimum version" anywhere in our HTTP API
- Requiring the header to be present in the request is harmful for API experimentation
    - We should issue a *strong* recommendation (like every other API) to lock your requested API version in "production"

### PR DESCRIPTION
**Pull request by amitlissack**
## overview

Add API version header to responses.

## changelog

Define an api version. This is an integer starting at 1.

Middleware that examines the `opentrons-version` header in request.  The api version used to respond to the request is chosen in this manner: 
- if the request version is an integer and <= server api version, we use that version
- otherwise we use the server version.

The chosen version is attached to the request.state object for consumption by handlers.
Finally, the response will contain `opentrons-version` that holds the version used to handle the request.

Rename `exceptions.py` to `errors.py` and put error related helpers in there.
 
## review requests

- [x] Version string is ok
- [x] Header name is ok
- [x] Location of api version constant is ok
- [x] This closes #4633. (The ticket discusses validating request version headers. Is that something we want to do now?)

## risk assessment

None

closes #4633

### CODE REVIEW

File: robot-server/robot_server/__init__.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# This is the version of the API supported by the server. The
+# application version is in package.json.
+API_VERSION = "1.0.0"
```
Comment (mcous) [inline robot-server/robot_server/__init__.py:3]: As per #4633, this should be a simple integer that counts up. Any breaking change (major bump) should involve path changes (e.g. `/v2/`), but hopefully we can avoid needing to do that for a long time

File: robot-server/robot_server/service/app.py [hunk 2]
```diff
@@ -57,11 +57,24 @@ async def on_startup():
...
 @app.on_event("shutdown")
 async def on_shutdown():
-    """App shutfown handler"""
+    """App shutdown handler"""
     s = await get_rpc_server()
     await s.on_shutdown()
 
 
+RESPONSE_HEADERS = {
+    'X-Opentrons-Media-Type': f'opentrons.robot-server.{API_VERSION}'
```
Comment (mcous) [inline robot-server/robot_server/service/app.py:66]: I updated the #4633 ticket with a few thoughts on this:

- `X-...` media types [have been deprecated](https://tools.ietf.org/html/rfc6648) in favor of naming headers in a way to inherintly avoid collision (e.g. prefixing all our custom headers with `Opentrons-`)
- String parsing is a pain for this
- I'm down to go something way more simple, like `'Opentrons-Version': API_VERSION`

**Comment by amitlissack** [comment]
> Sorry for not doing this sooner, but I just went in and updated the #4633 spec. I simplified the spec based pretty much entirely on what the [Stripe API does](https://stripe.com/docs/api) because I like that API a lot.
> 
> Also, #4633 contains two halves: requests and response. Since this PR only addresses the response half, *_it does not close the ticket as written_. We should either keep the ticket open or write a new one for the request side of things

I think a separate ticket makes more sense.

File: robot-server/robot_server/service/errors.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    )
+
+
+def consolidate_fastapi_response(all_exceptions: List[Dict[str, Any]]) -> str:
+    """
+    Consolidate the default fastAPI response so it can be returned as a string.
+    Default schema of fastAPI exception response is:
+    {
+        'loc': ('body',
+                '<outer_scope1>',
```
Comment (sfoster1) [inline robot-server/robot_server/service/errors.py:59]: is this precisely 2 outer scopes or is it (body, [0-n outer scopes], inner_param)

Comment (amitlissack) [reply robot-server/robot_server/service/errors.py:59]: It is the latter.

**Pull request merged**